sse2neon.h

```diff
@@ -1693,6 +1708,46 @@ FORCE_INLINE __m128 _mm_mul_ps(__m128 a, __m128 b)
         vmulq_f32(vreinterpretq_f32_m128(a), vreinterpretq_f32_m128(b)));
 }
 
+// Multiply the low unsigned 32-bit integers from each packed 64-bit element in
+// a and b, and store the unsigned 64-bit results in dst.
+//
+//   r0 :=  (uint32_t*)a0 * (uint32_t*)b0
+//   r1 :=  (uint32_t*)a2 * (uint32_t*)b2
+#if 1 /* C version */
+FORCE_INLINE __m128i _mm_mul_epu32(__m128i a, __m128i b)
+{
+    __m128i d;
+    vreinterpretq_nth_u64_m128i(d, 0) =
+        (uint64_t)(vreinterpretq_nth_u32_m128i(a, 0)) *
+        (uint64_t)(vreinterpretq_nth_u32_m128i(b, 0));
+    vreinterpretq_nth_u64_m128i(d, 1) =
+        (uint64_t)(vreinterpretq_nth_u32_m128i(a, 2)) *
+        (uint64_t)(vreinterpretq_nth_u32_m128i(b, 2));
+    return d;
+}
+#else /* Neon version */
+// Default to c version until casting can be sorted out on neon version.
+// (Otherwise requires compiling with -fpermissive) Also unclear whether neon
+// version actually performs better.
+FORCE_INLINE __m128i _mm_mul_epu32(__m128i a, __m128i b)
+{
+    // shuffle: 0, 1, 2, 3 -> 0, 2, 1, 3 */
+    __m128i const a_shuf =
+        *(__m128i *) (&vzip_u32(vget_low_u32(vreinterpretq_u32_m128i(a)),
+                                vget_high_u32(vreinterpretq_u32_m128i(a))));
+    __m128i const b_shuf =
+        *(__m128i *) (&vzip_u32(vget_low_u32(vreinterpretq_u32_m128i(b)),
+                                vget_high_u32(vreinterpretq_u32_m128i(b))));
+
+    // Multiply low word (32 bit) against low word (24 bit) and high word (32
+    // bit) against high word (32 bit). Pack both results (64 bit) into 128 bit
+    // register and return result.
+    return vreinterpretq_m128i_u64(
+        vmull_u32(vget_low_u32(vreinterpretq_u32_m128i(a_shuf)),
+                  vget_low_u32(vreinterpretq_u32_m128i(b_shuf))));
+}
+#endif
```

Comment:
Has the `NEON` version been tested with the unit-test?
I think the implementation is not correct.
If it isn't, you can leave the implementation empty.

Reply:
Yes the NEON version passes the unit tests for me. I've also run it on a semi complex codebase where it appears to work. 
Its just disabled by default because of the casting and because its unclear whether there are performance benefits - but I think it serves as a useful base for others in future.